engineapi/proxy.go

```diff
@@ -1,168 +1,429 @@
 package engineapi
 
 import (
+	"fmt"
 	"sync"
 	"time"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/net/context"
+	"google.golang.org/grpc/connectivity"
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
 	imtypes "github.com/longhorn/longhorn-instance-manager/pkg/types"
 	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
 
 	"github.com/longhorn/longhorn-manager/datastore"
 	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/util"
 )
 
 const (
-	ProxyPollInterval = 5 * time.Second
+	MaxProxyClientConcurrentInUse = 1000
+	DefaultProxyClientPerImage    = 3
+	MaxProxyClientPerImage        = 3 * DefaultProxyClientPerImage
+	ProxyPollInterval             = 5 * time.Second
+	ProxyClientMaxRetry           = 10
+	ProxyGCMaxRetry               = 3
 )
 
-type EngineClientProxy interface {
-	EngineClient
-	BackupTargetBinaryClient
-
-	IsGRPC() bool
-	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
-	Stop(string) error
-	Ping() error
-}
-
 type EngineClientProxyHandler struct {
-	Clients map[string]EngineClientProxy
+	proxyPools map[string]engineClientProxyPool
 
 	logger logrus.FieldLogger
+	ds     *datastore.DataStore
 
-	ds *datastore.DataStore
+	stopCh chan struct{}
 
 	lock *sync.RWMutex
+
+	resyncPeriod time.Duration
+}
+
+type engineClientProxyPool struct {
+	clients map[string]EngineClientProxy
+
+	instanceManagerName string
 }
 
-func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore) *EngineClientProxyHandler {
+func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore, resyncPeriod time.Duration, stopCh chan struct{}) *EngineClientProxyHandler {
 	return &EngineClientProxyHandler{
-		Clients: make(map[string]EngineClientProxy),
-		logger:  logger,
-		ds:      ds,
-		lock:    &sync.RWMutex{},
+		proxyPools:   make(map[string]engineClientProxyPool),
+		logger:       logger,
+		ds:           ds,
+		stopCh:       stopCh,
+		lock:         &sync.RWMutex{},
+		resyncPeriod: resyncPeriod,
```

Comment:
NIT: This actually means the resync period of a single proxy. Rename this to avoid misunderstanding.